server/fast_forward.go

```diff
@@ -0,0 +1,223 @@
+// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package server
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"net/http"
+	"regexp"
+	"strings"
+	"text/template"
+	"time"
+
+	"github.com/google/go-github/v39/github"
+	"github.com/mattermost/mattermost-mattermod/model"
+	"github.com/mattermost/mattermost-server/v5/mlog"
+)
+
+const (
+	cloudBranchName        = "heads/cloud"
+	mainBranchName         = "heads/master"
+	processRepo            = "mattermost-server"
+	minBackupIntervalHours = 5 * 24
+)
+
+var (
+	backupRegex = regexp.MustCompile(`cloud-20\d{2}-\d{2}-\d{2}-backup`)
+)
+
+type fastForwardResult struct {
+	Backup        []string
+	Skipped       []string
+	FastForwarded []string
+}
+
+// HasCloudFF is true if body contains "/cloud-ff"
+func (e *issueCommentEvent) HasCloudFF() bool {
+	return strings.Contains(strings.TrimSpace(e.Comment.GetBody()), "/cloud-ff")
+}
+
+func (s *Server) performFastForwardProcess(ctx context.Context, issue *model.Issue, comment, user string) (*fastForwardResult, error) {
+	if issue.State == model.StateClosed || issue.RepoName != processRepo { // we will perform this only on mm-server issues
+		return nil, nil
+	}
+
+	// Don't start process if the user is not a core committer
+	if !s.IsOrgMember(user) {
+		return nil, nil
+	}
+
+	// We are using a mutex for multi-deployment setups
+	mx := s.Store.Mutex()
+	if err := mx.Lock(ctx); err != nil {
+		return nil, err
+	}
+	defer func() {
+		if err := mx.Unlock(); err != nil {
+			mlog.Error("unable to release lock", mlog.Err(err))
+		}
+	}()
+
+	result := &fastForwardResult{}
+
+	// Check if the args are correct
+	command := getFFCommand(comment)
+	args := strings.Split(command, " ")
+	mlog.Info("Args", mlog.String("Args", comment))
+	var force bool
+	if len(args) >= 2 {
+		force = args[1] == "--force"
+	}
+
+	backupBranchName := cloudBranchName + "-" + time.Now().Format("2006-01-02") + "-backup"
+
+	for _, repo := range s.Config.CloudRepositories {
+		refs, _, err := s.GithubClient.Git.ListMatchingRefs(ctx, s.Config.Org, repo, &github.ReferenceListOptions{
```

Comment:
Out of scope for this PR, but do we need to worry about increasing amount of refs as time passes? Wondering if we should delete the older branches to keep the number of refs constant?